### ISSUE TITLE
[v2] Architecture for framabag

### ISSUE THREAD

**Issue by nicosomb** [wallabag/wallabag#1330]
### Sources

Today, Framabag is multi instances, one folder per user, with symlinks to a `core` folder which contains wallabag 1.x sources. 

Advantages:
* each user has his own database
* each user doesn't disturb other users if his database grows

About wallabag v2, I wonder if single instance wouldn't be a better solution. 

### Database

Today, each user has his own SQLite database. SQLite, it's cool, it's lite ... but with Postgres, growth opportunities seem better.

Disadvantage: one database for every user, latency problems?, each user can't export his own database (but json export will be possible), etc.

### Misc

* [Redis](http://redis.io/), to improve fetching of articles
* [Elasticsearch](https://www.elastic.co/products/elasticsearch), to improve our poor search engine. Elasticsearch is also able to replace our database and store all entries.

**Comment by tcitworld** [comment]
> our poor search engine

But, but... Someone told me our search engine was perfect...

Just kidding.

I'm in for the PostgreSQL : one database (or a little number if we reach too much users) for all users.

**Comment by j0k3r** [comment]
:+1: for Postgre
We can think about scaling the app later. I think we won't jump framabag on v2 right after the first release hein? :)

**Comment by nicosomb** [comment]
Don't worry, we'll migrate framabag with 2.0, not the alpha. And we have still lot of work about migration (scripts, etc). 

@FabienM any public opinion about redis /ES ? 



-------- Message d'origine --------
De : Jeremy Benoist <notifications@github.com>
Envoyé : 12 août 2015 22:01:30 UTC+02:00
À : wallabag/wallabag <wallabag@noreply.github.com>
Cc : "Nicolas Lœuillet" <nicolas@loeuillet.org>
Objet : Re: [wallabag] [v2] Architecture for framabag (#1330)

:+1: for Postgre
We can think about scaling the app later. I think we won't jump framabag on v2 right after the first release hein? :)

---

**Comment by FabienM** [comment]
Hi everybody, thank you for asking me :-)

**Postgres** is the best OSS DBMS for relational modeling and it can also works as a document-oriented database with advanced json features (not as much as mongo, but still powerful).
This choice is good, almost overkill since Wallabag model is really simple and not so relational, and the application does not need advanced transaction management and ACID capabilities. It is still a reasonnable choice.

I think that the SQLite solution will reach its limits sooner or later, as framabag is growing : hardware resources such as I/O and CPU will begin to suffer.

**Redis** is a good solution for storing key-value pairs but managing structured models may be a pain : one need to map the model to a "Hash" and customs indexes need to be maintained "at-hand", writing the same model in multiple cahe keys. So I would not suggest it as a main datastore.
Otherwise, the PUBSUB feature could be handful to create a "queue" of articles to fetch and SncRedisBundle allows to use Redis as Doctrine, Symfony and session cache, and it worth it.

**Elastic**~~Search~~ sounds obvious if you intend to implement advanced seach features. It's scalable in shards (but YAGNI) and there is a high probability that it will end up in your stack anyway.
Data are stored in Apache Lucene indexes so consistency and durabilty won't be a matter. 


To summarize : If you intend to use it as search engine someday, I would probably use it as a single DataStore for an application like Framabag, since PostgreSQL can sometimes be hard to install, manage & optimize.
If PostgreSQL already suits search features you want, just keep it.

Best regards,

Fabien

**Comment by nicosomb** [comment]
v2.wallabag.org uses postgres, and will use RabbitMQ and ES (so next version for "saas wallabag" will use them too).
